### PR TITLE
feat: libp2p relay and gossipsub server

### PR DESCRIPTION
**Pull request by Larkooo**

### CODE REVIEW

File: crates/torii/libp2p/src/client/mod.rs [hunk 1]
```diff
@@ -0,0 +1,212 @@
...
+                    ping: ping::Behaviour::new(ping::Config::default()),
+                }
+            })?
+            .build();
+
+        info!("Dialing relay: {:?}", relay_addr);
+        swarm.dial(relay_addr.parse::<Multiaddr>()?)?;
+
+        let (message_sender, message_receiver) = futures::channel::mpsc::channel(0);
+        let (command_sender, command_receiver) = futures::channel::mpsc::channel(0);
```
[the remark this answers is not in the document]
Comment (tarrencev) [reply crates/torii/libp2p/src/client/mod.rs:83]: I agree that they should all be handled. Afaiu that would still be the case with a buffer. With no buffer, the publisher will be blocked by the consumer

File: crates/torii/client/src/client/mod.rs [hunk 2]
```diff
@@ -34,6 +36,8 @@ pub struct Client {
     metadata: Arc<RwLock<WorldMetadata>>,
     /// The grpc client.
     inner: AsyncRwLock<torii_grpc::client::WorldClient>,
+    /// Libp2p client.
+    libp2p_client: torii_relay::client::RelayClient,
```
Comment (tarrencev) [inline crates/torii/client/src/client/mod.rs:40]:
```suggestion
    relay_client: torii_relay::client::RelayClient,
```

File: crates/torii/libp2p/src/client/mod.rs [hunk 3]
```diff
@@ -0,0 +1,217 @@
...
+                    }
+                },
+            }
+        }
+    }
+
+    fn subscribe(&mut self, room: &str) -> Result<bool, Error> {
+        let topic = IdentTopic::new(room);
+        let sub = self.swarm.behaviour_mut().gossipsub.subscribe(&topic)?;
+        // if sub {
```
Comment (tarrencev) [inline crates/torii/libp2p/src/client/mod.rs:193]: remove or?

File: crates/torii/libp2p/src/client/mod.rs [hunk 4]
```diff
@@ -0,0 +1,217 @@
...
+        //     self.topics.insert(room.to_string(), topic);
+        // }
+
+        Ok(sub)
+    }
+
+    fn unsubscribe(&mut self, room: &str) -> Result<bool, Error> {
+        let topic = IdentTopic::new(room);
+        let unsub = self.swarm.behaviour_mut().gossipsub.unsubscribe(&topic)?;
+        // if unsub {
```
Comment (tarrencev) [inline crates/torii/libp2p/src/client/mod.rs:203]: cleanup?

File: crates/torii/libp2p/src/client/mod.rs (hunk 3 above)
Comment (Larkooo) [reply crates/torii/libp2p/src/client/mod.rs:193]: Mhh do we want to keep track of the subscribed topics? They're hashed when subscribed to so we can't get the topic names we subscribed to, only the hash.

**Comment by Larkooo** [comment]
@glihm would be great if you could take a look at some of the event loop and channel stuff to see if anything looks odd. does taking commands from a channel make sense? that solves most of the borrow checker issues when running the client.

File: crates/torii/libp2p/src/server/mod.rs [hunk 5]
```diff
@@ -0,0 +1,228 @@
...
+                        .validation_mode(gossipsub::ValidationMode::Strict) // This sets the kind of message validation. The default is Strict (enforce message signing)
+                        .message_id_fn(message_id_fn) // content-address messages. No two messages of the same content will be propagated.
+                        .build()
+                        .map_err(|msg| io::Error::new(io::ErrorKind::Other, msg)).unwrap(); // Temporary hack because `build` does not return a proper `std::error::Error`.
+
+                Behaviour {
+                    relay: relay::Behaviour::new(key.public().to_peer_id(), Default::default()),
+                    ping: ping::Behaviour::new(ping::Config::new()),
+                    identify: identify::Behaviour::new(identify::Config::new(
+                        "/TODO/0.0.1".to_string(),
```
Comment (glihm) [inline crates/torii/libp2p/src/server/mod.rs:89]: Is this `TODO` not a problem for now? Should it be `torii-server/relay`?

File: crates/torii/libp2p/src/server/mod.rs [hunk 6]
```diff
@@ -0,0 +1,228 @@
...
+        let listen_addr_webrtc = Multiaddr::from(Ipv4Addr::UNSPECIFIED)
+            .with(Protocol::Udp(port_webrtc))
+            .with(Protocol::WebRTCDirect);
+        swarm.listen_on(listen_addr_webrtc.clone())?;
+
+        // Clients will send their messages to the "message" topic
+        // with a room name as the message data.
+        // and we will forward those messages to a specific room - in this case the topic
+        // along with the message data.
+        swarm.behaviour_mut().gossipsub.subscribe(&IdentTopic::new("message")).unwrap();
```
Comment (glihm) [inline crates/torii/libp2p/src/server/mod.rs:121]: This `"message"` topic may be relocated to a common variable for client and server.

**Review by glihm — COMMENTED** [review]
Awesome work @Larkooo. :rocket:

Some minor refacto comments to anticipate future PRs.

About the channels, it looks good to me. We've to keep in mind the way `unbounded` aborts the process (when out of memory), this to ensure we can identify a possible error cause in the future associated to that.

Would be happy to dive into the borrow checker issues you had to understand how you reached the channel. :+1:

**Review by glihm — COMMENTED** [review]
Thanks for the update and nice addition for the `CommandSender`.
Will you add the `run_libp2p` in this run, or is it a work for the next PR?

**Comment by Larkooo** [comment]
> Thanks for the update and nice addition for the `CommandSender`. Will you add the `run_libp2p` in this run, or is it a work for the next PR?

It has been added, pretty simple, just awaits the event_loop.run(). it's in torii-client. I think for now that's good enough

**Review by glihm — APPROVED** [review]

**Pull request merged**